### PR TITLE
ordering by size in clustrmsd

### LINKED ISSUE #955: change cluster IDs to reflect cluster sizes in clustrmsd
## **Desired feature/enhancement**
In the haddock2.X series we use to have the clustered numbered according to their size, with 1 being the largest cluster. 

## **Motivation**
This should be preserved here to avoid confusion

## **Description**
change the way cluster IDs are assigned, ordering them by the size of the clusters

### PR DESCRIPTION
**Pull request by mgiulini**
You are about to submit a new Pull Request. Before continuing make sure you read the [contributing guidelines](CONTRIBUTING.md) and that you comply with the following criteria:

- [x] You have sticked to Python. Please talk to us before adding other programming languages to HADDOCK3
- [ ] Your PR is about CNS
- [x] Your code is well documented: proper docstrings and explanatory comments for those tricky parts
- [x] You structured the code into small functions as much as possible. You can use classes if there is a (state) purpose
- [x] Your code follows our coding style
- [x] You wrote tests for the new code
- [x] `tox` tests pass. *Run `tox` command inside the repository folder*
- [x] `-test.cfg` examples execute without errors. *Inside `examples/` run `python run_tests.py -b`*
- [ ] PR does not add any dependencies, unless permission granted by the HADDOCK team
- [ ] PR does not break licensing
- [ ] Your PR is about writing documentation for already existing code :fire:
- [ ] Your PR is about writing tests for already existing code :godmode:

---

Closes #955 by ordering cluster by size in clustrmsd

### CODE REVIEW

File: tests/test_module_clustrmsd.py [hunk 1]
```diff
@@ -352,3 +353,20 @@ def test_iterate_min_population():
...
+def test_order_clusters():
+    """Test order_clusters function."""
+    cluster_arr = np.array([1, 1, 2, 3, 4])
+    obs_clusters, obs_cluster_arr = order_clusters(cluster_arr)
+    exp_clusters = [1, 2, 3, 4]
+    exp_cluster_arr = np.array([1, 1, 2, 3, 4])
+    assert obs_clusters == exp_clusters
+    assert (obs_cluster_arr == exp_cluster_arr).all()
+    # now with a less trivial cluster_arr
+    cluster_arr = np.array([3, 3, 2, 4, 3, 1, 3, 3, 1, 3, 4])
```
Comment (VGPReys) [inline tests/test_module_clustrmsd.py:367]: I have the feeling that cluster id `-1` is not tested

File: src/haddock/modules/analysis/clustrmsd/clustrmsd.py [hunk 2]
```diff
@@ -356,3 +356,39 @@ def write_clustrmsd_file(clusters, clt_dic, cluster_centers, score_dic, sorted_s
...
+    sorted_clusters = cluster_pops[0][sorted_indices]
+    
+    # delete -1 from sorted_clusters if present
+    sorted_clusters = sorted_clusters[sorted_clusters != -1]
+    clusters = []
+    for c in sorted_clusters:
+        cluster_arr[cluster_arr == c] = -c
+    for i, c in enumerate(sorted_clusters):
+        clusters.append(i + 1)
+        cluster_arr[cluster_arr == -c] = i + 1
```
Comment (VGPReys) [inline src/haddock/modules/analysis/clustrmsd/clustrmsd.py:393]:
```suggestion
    for c in sorted_clusters:
        cluster_arr[cluster_arr == c] = -c
    clusters = []
    for ci, c in enumerate(sorted_clusters, start=1):
        clusters.append(ci)
        cluster_arr[cluster_arr == -c] = ci
```

File: src/haddock/modules/analysis/clustrmsd/clustrmsd.py [hunk 3]
```diff
@@ -356,3 +356,39 @@ def write_clustrmsd_file(clusters, clt_dic, cluster_centers, score_dic, sorted_s
...
+    clusters : list
+        List of clusters.
+    
+    cluster_arr : np.ndarray
+        Array of clusters.
+    """
+    cluster_pops = np.unique(cluster_arr, return_counts=True)
+    
+    sorted_indices = np.argsort(-cluster_pops[1])
+    sorted_clusters = cluster_pops[0][sorted_indices]
```
Comment (VGPReys) [inline src/haddock/modules/analysis/clustrmsd/clustrmsd.py:384]:
```suggestion
    unique_clusters, cluster_counts = np.unique(cluster_arr, return_counts=True)
    
    sorted_indices = np.argsort(-cluster_counts)  # must use negative to sort ascending
    sorted_clusters = unique_clusters[sorted_indices]
```

**Review by VGPReys — DISMISSED** [review]

File: tests/test_module_clustrmsd.py (hunk 1 above)
Comment (mgiulini) [reply tests/test_module_clustrmsd.py:367]: good catch! I corrected the function and added the test, should be fine now

**Review by VGPReys — APPROVED** [review]

**Review by amjjbonvin — APPROVED** [review]
Approved - once the main branch has been merged into it

**Pull request merged**